Recast/Source/Recast.cpp

```diff
@@ -262,319 +260,284 @@ rcPolyMesh::~rcPolyMesh()
 
 rcPolyMeshDetail* rcAllocPolyMeshDetail()
 {
-	rcPolyMeshDetail* dmesh = (rcPolyMeshDetail*)rcAlloc(sizeof(rcPolyMeshDetail), RC_ALLOC_PERM);
-	memset(dmesh, 0, sizeof(rcPolyMeshDetail));
-	return dmesh;
+	return rcNew<rcPolyMeshDetail>(RC_ALLOC_PERM);
+}
+
+void rcFreePolyMeshDetail(rcPolyMeshDetail* detailMesh)
+{
+	rcDelete(detailMesh);
+}
+
+rcPolyMeshDetail::rcPolyMeshDetail()
+: meshes()
+, verts()
+, tris()
+, nmeshes()
+, nverts()
+, ntris()
+{
 }
 
-void rcFreePolyMeshDetail(rcPolyMeshDetail* dmesh)
+rcPolyMeshDetail::~rcPolyMeshDetail()
 {
-	if (!dmesh) return;
-	rcFree(dmesh->meshes);
-	rcFree(dmesh->verts);
-	rcFree(dmesh->tris);
-	rcFree(dmesh);
+	rcFree(meshes);
+	rcFree(verts);
+	rcFree(tris);
```

Comment:
This change breaks compatibility with code using older versions of RecastNavigation causing double free since previously there was no destructor and it was required to call `rcFree` for this fields from client code.

Reply:
I don't quite follow why this would cause a double-free.  This is just moving the calls to `rcFree` the poly mesh members from `rcFreePolyMeshDetail` to `~rcPolyMeshDetail`.  

It sounds like you might not be calling `rcFreePolyMeshDetail` and are instead making the same calls to `rcFree` yourself?  In that case, I don't think it's going to result in a double-free, since you'd be calling `rcFree` on the `rcPolyMeshDetail` instance, which bypasses the destructor.  

The only case I can imagine where this might cause a double-free is if you're calling the destructor explicitly _and_ are calling `rcFreePolyMeshDetail`.

Can you post some usage code that is now a double-free as a result of this change?

Regardless, any double-free issues here could be resolved by zero-ing these pointers after free, which is probably a good idea anyway.

Reply:
The problem is that when `rcPolyMeshDetail` is constructed without use of `rcAllocPolyMeshDetail`. It's not used because there is no need to allocate an object of 3 number and 3 pointers on a heap. And in general it's not a good idea to force users to allocate objects on a heap from the performance point of view.

The client code is open source. Basically a combination of [this](https://github.com/OpenMW/openmw/blob/3f6b33820a4b660285e4dc7d83f48e83a84d0a46/components/detournavigator/preparednavmeshdata.hpp#L18) and [this](https://github.com/OpenMW/openmw/blob/3f6b33820a4b660285e4dc7d83f48e83a84d0a46/components/detournavigator/preparednavmeshdata.cpp#L40) triggered the double free. [This](https://gitlab.com/OpenMW/openmw/-/merge_requests/2564) is how I solved the problem to make the code compatible with both versions of the recastnavigation.

> Regardless, any double-free issues here could be resolved by zero-ing these pointers after free, which is probably a good idea anyway.

I don't agree. Double free should not happen in the first place. If it happens there is a flaw in the code logic. Better to fix the flaw rather than hide it by zeroing pointer. Unfortunatelly in my case it's not possible because I need to keep compatibility with multiple recastnavigation versions.

Reply:
Thanks for the clarification.  

To recap your situation, it comes down to creating a `rcPolyMeshDetail` without calling `rcAllocPolyMeshDetail`, not calling `rcFreePolyMeshDetail`, and manually freeing the three array pointers in `rcPolyMeshDetail` in some custom routine.

To be clear, I'm not necessarily advocating that _just_ null-ing a freed pointer is a good enough solution.  I agree with you that double-frees should be avoided.  My point was more that zero-ing free'd pointers is just good code hygiene and something we should consider.

Moving these `rcFree` calls into the destructor of `rcPolyMeshDetail` makes your use case easier to implement and less prone to memory bugs.  To your point, we shouldn't force people to use the heap, and moving the `rcFree` calls to the destructor means you're not forced to implement them yourself if you're not using `rcFreePolyMeshDetail`, like you've done here.

For the sake of semver and maintaining perfect backwards compatibility, I think I'll move this specific change to the v2.0 branch and keep the old functionality in main.